File: Applied Course/4.Problem Solving/12.Problems on Backtracking/6.Word Search.py
# https://leetcode.com/problems/word-search/
class Solution:
    def pathExist(self,row,col,word):
        if len(word)==0:
            return True
        if row<0 or col<0 or row==self.ROWS or col==self.COLS or self.board[row][col]!=word[0]:
            return False
        ret=False
        self.board[row][col]='@'
        for rowoffset,coloffset in [(0,1),(1,0),(-1,0),(0,-1)]:
            ret=self.pathExist(row+rowoffset,col+coloffset,word[1:])
            if ret:
                break
        self.board[row][col]=word[0]
        return ret
    def exist(self, board: List[List[str]], word: str) -> bool:
        self.ROWS=len(board)
        self.COLS=len(board[0])
        self.board=board
        for i in range(self.ROWS):
            for j in range(self.COLS):
                if self.pathExist(i,j,word):
                    return True
        return False

File: Applied Course/4.Problem Solving/12.Problems on Backtracking/6.Word Search.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    bool solve(vector<vector<char>>& board, int r, int c, int R, int C, string word, int idx){
        if(word.size() == idx) 
            return true;
        if(r<0 || r>=R || c<0 || c>=C || board[r][c]=='$' || board[r][c]!=word[idx]) 
            return false;
        char ch = board[r][c];
        board[r][c] ='$';
        int ans=false;
        if(solve(board, r+1,c,R,C,word, idx+1) || solve(board, r-1,c,R,C,word, idx+1) || solve(board, r,c+1,R,C,word, idx+1) || solve(board, r,c-1,R,C,word, idx+1))
                ans = true;
        board[r][c] = ch;
        return ans;
    }
    bool exist(vector<vector<char>>& board, string word) {
        int ROW = board.size(), COL= board[0].size();
        if(ROW*COL< word.size()) 
            return 0;        
        for(int i=0;i<ROW;i++){
            for(int j=0;j<COL;j++){
                if(board[i][j] == word[0]){
                    if(solve(board, i,j, ROW, COL, word, 0)) 
                        return true;
                }
            }
        }
        return false;
    }
};
int main()
{
    // https://leetcode.com/problems/word-search
    return 0;
}